### PR TITLE
[Fix #7161] Fix `Style/SafeNavigation` cop to preserve comments

### LINKED ISSUE #7161: Style/SafeNavigation autocorrect removes comments
## Expected behavior

When rewriting code to replace `if` statements with safe navigation, I expect comments to remain intact.

## Actual behavior

Comments within the `if` statement are removed.

## Steps to reproduce the problem
```ruby
if wal
  # Description of what happens here
  wal.release(lock)
end
```
autocorrects to
```ruby
wal&.release(lock)
```
rather than
```ruby
# Description of what happens here
wal&.release(lock)
```

## RuboCop version
```
0.69.0 (using Parser 2.6.2.0, running on ruby 2.4.5 x86_64-darwin17)
```

### PR DESCRIPTION
**Pull request by tejasbubane**
Closes #7161.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: spec/rubocop/cop/style/safe_navigation_spec.rb [hunk 1]
```diff
@@ -933,6 +933,23 @@
...
+            end
+          RUBY
+
+          expected_source = <<~RUBY
+            # this is a comment
+            # another comment
+            #{variable}&.bar
+          RUBY
+          expect(new_source).to eq(expected_source)
+        end
```
Comment (koic) [inline spec/rubocop/cop/style/safe_navigation_spec.rb]: Hmm, indent seems to be broken when used in a method.

```diff
diff --git a/spec/rubocop/cop/style/safe_navigation_spec.rb b/spec/rubocop/cop/style/safe_navigation_spec.rb
index 2c142df19..10697d908 100644
--- a/spec/rubocop/cop/style/safe_navigation_spec.rb
+++ b/spec/rubocop/cop/style/safe_navigation_spec.rb
@@ -935,17 +935,21 @@ RSpec.describe RuboCop::Cop::Style::SafeNavigation, :config do

         it 'does not lose comments within if expression' do
           new_source = autocorrect_source(<<~RUBY)
-            if #{variable}
-              # this is a comment
-              # another comment
-              #{variable}.bar
+            def foo
+              if #{variable}
+                # this is a comment
+                # another comment
+                #{variable}.bar
+              end
             end
           RUBY

           expected_source = <<~RUBY
-            # this is a comment
-            # another comment
-            #{variable}&.bar
+            def foo
+              # this is a comment
+              # another comment
+              #{variable}&.bar
+            end
           RUBY
           expect(new_source).to eq(expected_source)
         end
```

```console
% bundle exec rspec spec/rubocop/cop/style/safe_navigation_spec.rb

(snip)

Failures:

  1) RuboCop::Cop::Style::SafeNavigation auto-correct behaves like all variable types if expression does not lose comments within if expression
     Failure/Error: expect(new_source).to eq(expected_source)

       expected: "def foo\n  # this is a comment\n  # another comment\n  FOO::BAR&.bar\nend\n"
            got: "def foo\n  # this is a comment\n# another comment\nFOO::BAR&.bar\nend\n"

       (compared using ==)

       Diff:
       @@ -1,6 +1,6 @@
        def foo
          # this is a comment
       -  # another comment
       -  FOO::BAR&.bar
       +# another comment
       +FOO::BAR&.bar
        end

     Shared Example Group: "all variable types" called from ./spec/rubocop/cop/style/safe_navigation_spec.rb:1193
     # ./spec/rubocop/cop/style/safe_navigation_spec.rb:954:in `block (5 levels) in <top (required)>'
```

Comment (jonas054) [reply spec/rubocop/cop/style/safe_navigation_spec.rb]: I think that's acceptable. Some of the `Layout` cops will fix the indentation in the auto-correct loop.

Comment (koic) [reply spec/rubocop/cop/style/safe_navigation_spec.rb]: I get it. @jonas054's opinion is excellent.

**Comment by jonas054** [comment]
@tejasbubane I think this is ready to merge if you just fix the conflicts.

**Comment by tejasbubane** [comment]
@jonas054 Fixed merge conflicts. I think the build failure is unrelated.

File: lib/rubocop/cop/style/safe_navigation.rb [hunk 2]
```diff
@@ -120,13 +120,22 @@ def autocorrect(node)
...
         end
 
         private
 
+        def handle_comments(corrector, method_call)
+          return if processed_source.comments.empty?
+
+          comments = processed_source.comments.map(&:text).join("\n")
+          corrector.insert_before(method_call.loc.expression,
+                                  comments + "\n")
```
Comment (buehmann) [inline lib/rubocop/cop/style/safe_navigation.rb]: This seems to insert all the comments from the file, not matter where they are located.

Comment (jonas054) [reply lib/rubocop/cop/style/safe_navigation.rb]: You're absolutely right! This is something we all missed. Thanks for the heads-up.

@tejasbubane Can you take a look?

Comment (jonas054) [reply lib/rubocop/cop/style/safe_navigation.rb]: The problem is reported in #7397 and fixed in #7420.

**Pull request merged**